ion/binarywriter.go

```diff
@@ -251,22 +249,51 @@ func (w *binaryWriter) WriteTimestamp(val Timestamp) error {
 	return w.writeValue("Writer.WriteTimestamp", buf)
 }
 
-// WriteSymbol writes a symbol value.
-func (w *binaryWriter) WriteSymbol(val string) error {
-	id, err := w.resolve("Writer.WriteSymbol", val)
-	if err != nil {
-		w.err = err
-		return err
+// WriteSymbol writes a symbol value given a SymbolToken.
+func (w *binaryWriter) WriteSymbol(val SymbolToken) error {
+	var id uint64
+	if val.LocalSID != SymbolIDUnknown {
+		id = uint64(val.LocalSID)
+	} else if val.Text != nil {
+		text := *val.Text
+		if _, ok := symbolIdentifier(text); ok {
+			// Wrap text value in single quotes if the symbol's text is a symbol identifier
+			// (ie. of form $n for some integer n)
+			// This is done to distinguish from actual symbol table mappings.
+			text = fmt.Sprintf("'%v'", text)
```

Comment:
I'm not sure I understand the behavior that's intended here. Why are we wrapping `$37` in single quotes?

If I pass a `SymbolToken` with `LocalSID=SymbolIDUnknown` and `Text="$37"`, I would expect `WriteSymbol` to try to resolve the text literal `$37` (unmodified) in the symbol table. If I'd wanted to write SID 37, I would have set `LocalSID=37` in the `SymbolToken`, right?

Reply:
We have roundtrip tests (such as `ion-tests/iontestdata/good/symbols.ion` with the test case `'$99'`) that fail without any internal logic wrapping `$99` in quotes.

When the text reader reads `'$99'`, it produces the string `"$99"` (note the lack of single quotes) and then creates a SymbolToken using that string value. The absence of single quotes creates confusion further downstream about whether we are dealing with an actual symbol table mapping vs a non-mapping SymbolToken with this text value.

I moved the single quote wrapping logic over to `textReader::nextBeforeTypeAnnotations` instead of `binaryWriter::WriteSymbol` and `textWriter::WriteSymbol`.

Reply:
> such as ion-tests/iontestdata/good/symbols.ion with the test case '$99'

Here's [a link to this test case](https://github.com/amzn/ion-tests/blob/034503caa56683f086edb1cc05a1b6207c1b18af/iontestdata/good/symbols.ion#L26) for anyone who's interested.

> When the text reader reads '$99', it produces the string "$99" (note the lack of single quotes) and then creates a SymbolToken using that string value.

This sounds like the correct behavior to me.

> The absence of single quotes creates confusion further downstream about whether we are dealing with an actual symbol table mapping vs a non-mapping SymbolToken with this text value.

This is something we need to sort out; I don't think there's any writer API where the golang string `$99` should be reinterpreted as anything other than a string literal.

Can you link to the function(s) that are causing the problems? Why is the reinterpretation happening at all?

Reply:
I don't remember off the top of my head where specifically to go looking, but the root cause on this one is likely me. :)

At some point back before the introduction of `SymbolToken`, if the binary reader hit a symbol ID it couldn't resolve (or the text reader hit an unquoted ${n}), they'd return the string `${id}` instead of an error. The writers correspondingly mapped `WriteSymbol("$123")` to being a symbol with that ID, bypassing the symbol table. That came at the expense of being able to read/write an _actual_ symbol whose text is `${n}`, but that wasn't a thing I thought I would ever need to do (😬😅).

Now that everything at the reader/writer level is converted to SymbolTokens, any remnants of that questionable decision can and should be removed.

Reply:
This makes total sense. Thanks for the added context!

Reply:
Alright I reworked this so the Symbol Token's text value is the unquoted `$99`. 
However when the text writer is writing out the symbol in textutils.go:writeSymbol(), we still need to wrap this value in single quotes.